### PR TITLE
8343067: C2: revisit constant-offset AddP chains after successful input idealizations

### PR DESCRIPTION
**Pull request by robcasloz**
This changeset re-adds a constant-offset AddP node (`u`) to C2's IGVN worklist when its address is given by another AddP node (`use`) whose offset has changed. This makes it possible for `AddPNode::Ideal` to flatten the address computation in cases where the offset of the latter (`use->in(AddPNode::Offset)`) is found to be constant during IGVN:

idealization

The end result is the generation of fewer explicit address computation instructions.

##### Functionality

- tier1-5 (linux-x64, windows-x64, macosx-x64, linux-aarch64, macosx-aarch64; release and debug mode).

##### Performance

- Tested performance on a set of standard benchmark suites (DaCapo, SPECjbb2015, SPECjvm2008). No significant change was observed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343067](https://bugs.openjdk.org/browse/JDK-8343067): C2: revisit constant-offset AddP chains after successful input idealizations (**Enhancement** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21898/head:pull/21898` \
`$ git checkout pull/21898`

Update a local copy of the PR: \
`$ git checkout pull/21898` \
`$ git pull https://git.openjdk.org/jdk.git pull/21898/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21898`

View PR using the GUI difftool: \
`$ git pr show -t 21898`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21898.diff">https://git.openjdk.org/jdk/pull/21898.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21898#issuecomment-2457760229)
</details>

### CODE REVIEW

File: src/hotspot/share/opto/phaseX.cpp [hunk 1]
```diff
@@ -1634,12 +1635,19 @@ void PhaseIterGVN::add_users_of_use_to_worklist(Node* n, Node* use, Unique_Node_
...
+  // - check Stores for loop invariant, and
+  // - if the changed input is the offset, check constant-offset AddP users for
+  //   address expression flattening.
+  if (use_op == Op_AddP) {
     for (DUIterator_Fast i2max, i2 = use->fast_outs(i2max); i2 < i2max; i2++) {
       Node* u = use->fast_out(i2);
-      if (u->is_Mem())
+      if (u->is_Mem()) {
         worklist.push(u);
+      } else if (n == use->in(AddPNode::Offset) &&
```
Comment (vnkozlov) [inline src/hotspot/share/opto/phaseX.cpp:1647]: `n == use->in(AddPNode::Offset)` result can be saved outside loop in local var.

Comment (robcasloz) [reply src/hotspot/share/opto/phaseX.cpp]: Thanks, done (commit deb7c4e1).

**Review by vnkozlov — APPROVED** [review]
Good.

**Review by iwanowww — APPROVED** [review]
Looks good.

**Comment by robcasloz** [comment]
Thanks for reviewing @vnkozlov and @iwanowww!

**Pull request closed without merging**